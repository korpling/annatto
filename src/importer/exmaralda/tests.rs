use std::{env, path::Path, sync::mpsc};

use graphannis::update::GraphUpdate;
use insta::{assert_display_snapshot, assert_snapshot};
use itertools::Itertools;

use crate::{
    importer::Importer, progress::ProgressReporter, util::import_as_graphml_string_2,
    workflow::StatusMessage, Module,
};

use super::ImportEXMARaLDA;

#[test]
fn test_exb_fail_for_timeline() {
    let import = ImportEXMARaLDA::default();
    let import_path = "./tests/data/import/exmaralda/fail-corrupt_timeline/import/";
    let (sender, _receiver) = mpsc::channel();
    let r = import.import_corpus(Path::new(import_path), import.step_id(None), Some(sender));
    assert!(r.is_err());
    assert_display_snapshot!(r.err().unwrap());
    let document_path = "./tests/data/import/exmaralda/fail-corrupt_timeline/import/test_doc.exb";
    let mut u = GraphUpdate::default();
    assert!(import
        .import_document(
            "import/test_doc",
            Path::new(document_path),
            &mut u,
            &ProgressReporter::new(None, import.step_id(None), 1).unwrap(),
            &None,
        )
        .is_err());
}

#[test]
fn test_exb_fail_for_no_category() {
    let import = ImportEXMARaLDA::default();
    let import_path = "./tests/data/import/exmaralda/fail-no_category/";
    let (sender, _receiver) = mpsc::channel();
    let r = import.import_corpus(Path::new(import_path), import.step_id(None), Some(sender));
    assert!(r.is_err());
    assert_display_snapshot!(r.err().unwrap());
    let document_path = Path::new(import_path).join("test_doc.exb");
    let mut u = GraphUpdate::default();
    assert!(import
        .import_document(
            "fail-no_category/test_doc",
            document_path.as_path(),
            &mut u,
            &ProgressReporter::new(None, import.step_id(None), 1).unwrap(),
            &None
        )
        .is_err());
}

#[test]
fn test_exb_fail_for_no_speaker() {
    let import = ImportEXMARaLDA::default();
    let import_path = "./tests/data/import/exmaralda/fail-no_speaker/";
    let (sender, _receiver) = mpsc::channel();
    let r = import.import_corpus(Path::new(import_path), import.step_id(None), Some(sender));
    assert!(r.is_err());
    assert_display_snapshot!(r.err().unwrap());
    let document_path = Path::new(import_path).join("test_doc.exb");
    let mut u = GraphUpdate::default();
    assert!(import
        .import_document(
            "fail-no_speaker/test_doc",
            document_path.as_path(),
            &mut u,
            &ProgressReporter::new(None, import.step_id(None), 1).unwrap(),
            &None
        )
        .is_err());
}

#[test]
fn test_exb_fail_for_undefined_speaker() {
    let import = ImportEXMARaLDA::default();
    let import_path = "./tests/data/import/exmaralda/fail-undefined_speaker/";
    let (sender, _receiver) = mpsc::channel();
    let r = import.import_corpus(Path::new(import_path), import.step_id(None), Some(sender));
    assert!(r.is_err());
    assert_display_snapshot!(r.err().unwrap());
    let document_path = Path::new(import_path).join("test_doc.exb");
    let mut u = GraphUpdate::default();
    assert!(import
        .import_document(
            "fail-undefined_speaker/test_doc",
            document_path.as_path(),
            &mut u,
            &ProgressReporter::new(None, import.step_id(None), 1).unwrap(),
            &None
        )
        .is_err());
}

#[test]
fn test_exb_fail_for_unknown_tli() {
    let import = ImportEXMARaLDA::default();
    let import_path = "./tests/data/import/exmaralda/fail-unknown_tli/";
    let (sender, _receiver) = mpsc::channel();
    let r = import.import_corpus(Path::new(import_path), import.step_id(None), Some(sender));
    assert!(r.is_err());
    assert_display_snapshot!(r.err().unwrap());
    let document_path = Path::new(import_path).join("test_doc.exb");
    let mut u = GraphUpdate::default();
    assert!(import
        .import_document(
            "fail-unknown_tli/test_doc",
            document_path.as_path(),
            &mut u,
            &ProgressReporter::new(None, import.step_id(None), 1).unwrap(),
            &None
        )
        .is_err());
}

#[test]
fn test_exb_fail_for_bad_timevalue() {
    let import = ImportEXMARaLDA::default();
    let import_path = "./tests/data/import/exmaralda/fail-bad_timevalue/";
    let (sender, _receiver) = mpsc::channel();
    let r = import.import_corpus(Path::new(import_path), import.step_id(None), Some(sender));
    assert!(r.is_err());
    assert_display_snapshot!(r.err().unwrap());
    let document_path = Path::new(import_path).join("test_doc.exb");
    let mut u = GraphUpdate::default();
    assert!(import
        .import_document(
            "fail-bad_timevalue/test_doc",
            document_path.as_path(),
            &mut u,
            &ProgressReporter::new(None, import.step_id(None), 1).unwrap(),
            &None
        )
        .is_err());
}

#[test]
fn test_exb_fail_no_start_no_end() {
    let import = ImportEXMARaLDA::default();
    let import_path = "./tests/data/import/exmaralda/fail-no_start_no_end/";
    let (sender, _receiver) = mpsc::channel();
    let r = import.import_corpus(Path::new(import_path), import.step_id(None), Some(sender));
    assert!(r.is_err());
    assert_display_snapshot!(r.err().unwrap());
}

#[test]
fn test_fail_invalid() {
    let import = ImportEXMARaLDA::default();
    let import_path = "./tests/data/import/exmaralda/fail-invalid/import/";
    let (sender, _receiver) = mpsc::channel();
    let r = import.import_corpus(Path::new(import_path), import.step_id(None), Some(sender));
    assert!(r.is_err());
    assert_display_snapshot!(r.err().unwrap());
    let document_path = "./tests/data/import/exmaralda/fail-invalid/import/test_doc_invalid.exb";
    let mut u = GraphUpdate::default();
    assert!(import
        .import_document(
            "import/test_doc_invalid",
            Path::new(document_path),
            &mut u,
            &ProgressReporter::new(None, import.step_id(None), 1).unwrap(),
            &None
        )
        .is_err());
}

#[test]
fn test_exb_on_disk() {
    let r = test_exb("./tests/data/import/exmaralda/clean/import/", 0);
    assert_eq!(r.is_ok(), true, "Probing core test result {:?}", r);
}

#[test]
fn test_exb_broken_audio_on_disk() {
    let r = test_exb("./tests/data/import/exmaralda/broken_audio/import/", 1);
    assert_eq!(r.is_ok(), true, "Probing core test result {:?}", r);
}

#[test]
fn test_exb_pass_missing_type_attr_on_disk() {
    let r = test_exb("./tests/data/import/exmaralda/pass-no_tier_type/import/", 9);
    assert_eq!(r.is_ok(), true, "Probing core test result {:?}", r);
}

fn test_exb(
    import_path: &str,
    expected_warnings_count: usize,
) -> Result<(), Box<dyn std::error::Error>> {
    let (sender, receiver) = mpsc::channel();
    let actual = import_as_graphml_string_2(
        ImportEXMARaLDA::default(),
        env::current_dir()?.join(import_path), // IMPORTANT: test with absolute paths, this is what Annatto does at runtime
        None,
        true,
        Some(sender),
    )?;
    assert_snapshot!(actual.split('\n').sorted().join("\n")); // ordering lines alphabetically for now should fix an issue with the volatile order of elements in graphml snapshots
    let warnings = receiver
        .into_iter()
        .filter(|m| matches!(m, StatusMessage::Warning(..)))
        .collect_vec();
    assert_eq!(
        expected_warnings_count,
        warnings.len(),
        "Unexpected amount of warnings: {:?}",
        warnings
    );
    Ok(())
}
