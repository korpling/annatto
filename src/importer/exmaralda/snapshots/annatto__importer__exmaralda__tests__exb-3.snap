---
source: src/importer/exmaralda/tests.rs
expression: "actual.split('\\n').sorted().join(\"\\n\")"
---
            <data key="k0">dipl</data>
            <data key="k1">norm</data>
            <data key="k10">1</data>
            <data key="k11">0-1</data>
            <data key="k11">0-1</data>
            <data key="k11">0-2</data>
            <data key="k11">0-5</data>
            <data key="k11">1-2</data>
            <data key="k11">1-2</data>
            <data key="k11">2-3</data>
            <data key="k11">2-3</data>
            <data key="k11">2-3</data>
            <data key="k11">3-4</data>
            <data key="k11">3-5</data>
            <data key="k11">3-5</data>
            <data key="k11">4-5</data>
            <data key="k12"> </data>
            <data key="k12"> </data>
            <data key="k12"> </data>
            <data key="k12"> </data>
            <data key="k12"> </data>
            <data key="k12"> </data>
            <data key="k12">I&apos;m</data>
            <data key="k12">I</data>
            <data key="k12">New York</data>
            <data key="k12">New</data>
            <data key="k12">York</data>
            <data key="k12">am</data>
            <data key="k12">in</data>
            <data key="k12">in</data>
            <data key="k2">I&apos;m</data>
            <data key="k2">New</data>
            <data key="k2">York</data>
            <data key="k2">in</data>
            <data key="k3">test_doc</data>
            <data key="k4">tests/data/import/exmaralda/pass-no_tier_type/import/exmaralda/test_file.wav</data>
            <data key="k5">dipl</data>
            <data key="k5">dipl</data>
            <data key="k5">dipl</data>
            <data key="k5">dipl</data>
            <data key="k5">dipl</data>
            <data key="k5">norm</data>
            <data key="k5">norm</data>
            <data key="k5">norm</data>
            <data key="k5">norm</data>
            <data key="k5">norm</data>
            <data key="k5">norm</data>
            <data key="k5">norm</data>
            <data key="k5">norm</data>
            <data key="k6">I</data>
            <data key="k6">New York</data>
            <data key="k6">be</data>
            <data key="k6">in</data>
            <data key="k7">corpus</data>
            <data key="k7">corpus</data>
            <data key="k7">corpus</data>
            <data key="k7">file</data>
            <data key="k7">node</data>
            <data key="k7">node</data>
            <data key="k7">node</data>
            <data key="k7">node</data>
            <data key="k7">node</data>
            <data key="k7">node</data>
            <data key="k7">node</data>
            <data key="k7">node</data>
            <data key="k7">node</data>
            <data key="k7">node</data>
            <data key="k7">node</data>
            <data key="k7">node</data>
            <data key="k7">node</data>
            <data key="k7">node</data>
            <data key="k7">node</data>
            <data key="k7">node</data>
            <data key="k7">node</data>
            <data key="k7">node</data>
            <data key="k7">node</data>
            <data key="k8">I</data>
            <data key="k8">New York</data>
            <data key="k8">am</data>
            <data key="k8">in</data>
            <data key="k9">ADP</data>
            <data key="k9">ADP</data>
            <data key="k9">PRON</data>
            <data key="k9">VERB</data>
        </edge>
        </edge>
        </edge>
        </edge>
        </edge>
        </edge>
        </edge>
        </edge>
        </edge>
        </edge>
        </edge>
        </edge>
        </edge>
        </edge>
        </edge>
        </edge>
        </edge>
        </edge>
        </edge>
        </edge>
        </edge>
        </edge>
        </edge>
        </edge>
        </edge>
        </edge>
        </edge>
        </edge>
        </edge>
        </edge>
        </edge>
        </edge>
        </edge>
        </edge>
        </edge>
        </edge>
        </edge>
        </edge>
        </edge>
        </edge>
        </node>
        </node>
        </node>
        </node>
        </node>
        </node>
        </node>
        </node>
        </node>
        </node>
        </node>
        </node>
        </node>
        </node>
        </node>
        </node>
        </node>
        </node>
        </node>
        </node>
        </node>
        </node>
        </node>
        <edge id="e0" source="import/exmaralda/test_doc#t_dipl_T286-T1" target="import/exmaralda/test_doc#T286" label="Coverage/annis/">
        <edge id="e1" source="import/exmaralda/test_doc#t_dipl_T286-T1" target="import/exmaralda/test_doc#T0" label="Coverage/annis/">
        <edge id="e10" source="import/exmaralda/test_doc#a_dipl_T286-T4" target="import/exmaralda/test_doc#T286" label="Coverage/annis/">
        <edge id="e11" source="import/exmaralda/test_doc#a_dipl_T286-T4" target="import/exmaralda/test_doc#T0" label="Coverage/annis/">
        <edge id="e12" source="import/exmaralda/test_doc#a_dipl_T286-T4" target="import/exmaralda/test_doc#T1" label="Coverage/annis/">
        <edge id="e13" source="import/exmaralda/test_doc#a_dipl_T286-T4" target="import/exmaralda/test_doc#T2" label="Coverage/annis/">
        <edge id="e14" source="import/exmaralda/test_doc#a_dipl_T286-T4" target="import/exmaralda/test_doc#T3" label="Coverage/annis/">
        <edge id="e15" source="import/exmaralda/test_doc#a_norm_T286-T0" target="import/exmaralda/test_doc#T286" label="Coverage/annis/">
        <edge id="e16" source="import/exmaralda/test_doc#a_norm_T0-T1" target="import/exmaralda/test_doc#T0" label="Coverage/annis/">
        <edge id="e17" source="import/exmaralda/test_doc#a_norm_T1-T2" target="import/exmaralda/test_doc#T1" label="Coverage/annis/">
        <edge id="e18" source="import/exmaralda/test_doc#a_norm_T2-T4" target="import/exmaralda/test_doc#T2" label="Coverage/annis/">
        <edge id="e19" source="import/exmaralda/test_doc#a_norm_T2-T4" target="import/exmaralda/test_doc#T3" label="Coverage/annis/">
        <edge id="e2" source="import/exmaralda/test_doc#t_dipl_T1-T2" target="import/exmaralda/test_doc#T1" label="Coverage/annis/">
        <edge id="e20" source="import/exmaralda/test_doc#T3" target="import/exmaralda/test_doc#T4" label="Ordering/annis/">
        <edge id="e21" source="import/exmaralda/test_doc#T2" target="import/exmaralda/test_doc#T3" label="Ordering/annis/">
        <edge id="e22" source="import/exmaralda/test_doc#T1" target="import/exmaralda/test_doc#T2" label="Ordering/annis/">
        <edge id="e23" source="import/exmaralda/test_doc#T0" target="import/exmaralda/test_doc#T1" label="Ordering/annis/">
        <edge id="e24" source="import/exmaralda/test_doc#T286" target="import/exmaralda/test_doc#T0" label="Ordering/annis/">
        <edge id="e25" source="import/exmaralda/test_doc#t_dipl_T286-T1" target="import/exmaralda/test_doc#t_dipl_T1-T2" label="Ordering/annis/dipl">
        <edge id="e26" source="import/exmaralda/test_doc#t_dipl_T1-T2" target="import/exmaralda/test_doc#t_dipl_T2-T3" label="Ordering/annis/dipl">
        <edge id="e27" source="import/exmaralda/test_doc#t_dipl_T2-T3" target="import/exmaralda/test_doc#t_dipl_T3-T4" label="Ordering/annis/dipl">
        <edge id="e28" source="import/exmaralda/test_doc#t_norm_T286-T0" target="import/exmaralda/test_doc#t_norm_T0-T1" label="Ordering/annis/norm">
        <edge id="e29" source="import/exmaralda/test_doc#t_norm_T0-T1" target="import/exmaralda/test_doc#t_norm_T1-T2" label="Ordering/annis/norm">
        <edge id="e3" source="import/exmaralda/test_doc#t_dipl_T2-T3" target="import/exmaralda/test_doc#T2" label="Coverage/annis/">
        <edge id="e30" source="import/exmaralda/test_doc#t_norm_T1-T2" target="import/exmaralda/test_doc#t_norm_T2-T4" label="Ordering/annis/norm">
        <edge id="e31" source="import/exmaralda" target="import" label="PartOf/annis/">
        <edge id="e32" source="import/exmaralda/test_doc" target="import/exmaralda" label="PartOf/annis/">
        <edge id="e33" source="import/exmaralda/test_file.wav" target="import/exmaralda/test_doc" label="PartOf/annis/">
        <edge id="e34" source="import/exmaralda/test_doc#T286" target="import/exmaralda/test_doc" label="PartOf/annis/">
        <edge id="e35" source="import/exmaralda/test_doc#T0" target="import/exmaralda/test_doc" label="PartOf/annis/">
        <edge id="e36" source="import/exmaralda/test_doc#T1" target="import/exmaralda/test_doc" label="PartOf/annis/">
        <edge id="e37" source="import/exmaralda/test_doc#T2" target="import/exmaralda/test_doc" label="PartOf/annis/">
        <edge id="e38" source="import/exmaralda/test_doc#T3" target="import/exmaralda/test_doc" label="PartOf/annis/">
        <edge id="e39" source="import/exmaralda/test_doc#T4" target="import/exmaralda/test_doc" label="PartOf/annis/">
        <edge id="e4" source="import/exmaralda/test_doc#t_dipl_T3-T4" target="import/exmaralda/test_doc#T3" label="Coverage/annis/">
        <edge id="e5" source="import/exmaralda/test_doc#t_norm_T286-T0" target="import/exmaralda/test_doc#T286" label="Coverage/annis/">
        <edge id="e6" source="import/exmaralda/test_doc#t_norm_T0-T1" target="import/exmaralda/test_doc#T0" label="Coverage/annis/">
        <edge id="e7" source="import/exmaralda/test_doc#t_norm_T1-T2" target="import/exmaralda/test_doc#T1" label="Coverage/annis/">
        <edge id="e8" source="import/exmaralda/test_doc#t_norm_T2-T4" target="import/exmaralda/test_doc#T2" label="Coverage/annis/">
        <edge id="e9" source="import/exmaralda/test_doc#t_norm_T2-T4" target="import/exmaralda/test_doc#T3" label="Coverage/annis/">
        <node id="import">
        <node id="import/exmaralda">
        <node id="import/exmaralda/test_doc">
        <node id="import/exmaralda/test_doc#T0">
        <node id="import/exmaralda/test_doc#T1">
        <node id="import/exmaralda/test_doc#T2">
        <node id="import/exmaralda/test_doc#T286">
        <node id="import/exmaralda/test_doc#T3">
        <node id="import/exmaralda/test_doc#T4">
        <node id="import/exmaralda/test_doc#a_dipl_T286-T4">
        <node id="import/exmaralda/test_doc#a_norm_T0-T1">
        <node id="import/exmaralda/test_doc#a_norm_T1-T2">
        <node id="import/exmaralda/test_doc#a_norm_T2-T4">
        <node id="import/exmaralda/test_doc#a_norm_T286-T0">
        <node id="import/exmaralda/test_doc#t_dipl_T1-T2">
        <node id="import/exmaralda/test_doc#t_dipl_T2-T3">
        <node id="import/exmaralda/test_doc#t_dipl_T286-T1">
        <node id="import/exmaralda/test_doc#t_dipl_T3-T4">
        <node id="import/exmaralda/test_doc#t_norm_T0-T1">
        <node id="import/exmaralda/test_doc#t_norm_T1-T2">
        <node id="import/exmaralda/test_doc#t_norm_T2-T4">
        <node id="import/exmaralda/test_doc#t_norm_T286-T0">
        <node id="import/exmaralda/test_file.wav">
    </graph>
    <graph edgedefault="directed" parse.order="nodesfirst" parse.nodeids="free" parse.edgeids="canonical">
    <key id="k0" for="node" attr.name="dipl::abbreviation" attr.type="string"/>
    <key id="k1" for="node" attr.name="norm::abbreviation" attr.type="string"/>
    <key id="k10" for="node" attr.name="dipl::sentence" attr.type="string"/>
    <key id="k11" for="node" attr.name="annis::time" attr.type="string"/>
    <key id="k12" for="node" attr.name="annis::tok" attr.type="string"/>
    <key id="k2" for="node" attr.name="dipl::dipl" attr.type="string"/>
    <key id="k3" for="node" attr.name="annis::doc" attr.type="string"/>
    <key id="k4" for="node" attr.name="annis::file" attr.type="string"/>
    <key id="k5" for="node" attr.name="annis::layer" attr.type="string"/>
    <key id="k6" for="node" attr.name="norm::lemma" attr.type="string"/>
    <key id="k7" for="node" attr.name="annis::node_type" attr.type="string"/>
    <key id="k8" for="node" attr.name="norm::norm" attr.type="string"/>
    <key id="k9" for="node" attr.name="norm::pos" attr.type="string"/>
</graphml>
<?xml version="1.0" encoding="UTF-8"?>
<graphml>
